crates/ruff_python_semantic/src/context.rs

```diff
@@ -435,4 +444,22 @@ impl<'a> Context<'a> {
         }
         exceptions
     }
+
+    /// Generate a [`Snapshot`] of the current context.
+    pub fn snapshot(&self) -> Snapshot {
+        Snapshot {
+            scope_id: self.scope_id,
+            stmt_id: self.stmt_id,
+            in_annotation: self.in_annotation,
+            in_type_checking_block: self.in_type_checking_block,
+        }
+    }
+
+    /// Restore the context to the given [`Snapshot`].
+    pub fn restore(&mut self, snapshot: Snapshot) {
+        self.scope_id = snapshot.scope_id;
```

Comment:
Nit: It could make sense to destructure the snapshot here so that rust will give us compile errors when adding new fields